Generic type inference and intersection types in function parameters

**TypeScript Version:** 

$ ./node_modules/.bin/tsc --version
Version 1.8.7

**Code**

```ts
// A self-contained demonstration of the problem follows...
const parameterFn = (props:{store:string}) => alert(props.store)
const brokenFunction = <OwnProps>(f: (p: {dispatch: number} & OwnProps) => void) => (o: OwnProps) => o
export const Form3 = brokenFunction(parameterFn)({store: "hello"})
```

**Expected behavior:**

`OwnProps` is inferred as `{store: string}`. Therefore, the code compiles; `parameterFn` asks for less information than it's given.

**Actual behavior:** 
```
testing.ts(4,37): error TS2345: Argument of type '(props: { store: string; }) => void' is not assignable to parameter of type '(p: { dispatch: number; } & {}) => void'.
  Types of parameters 'props' and 'p' are incompatible.
    Type '{ store: string; }' is not assignable to type '{ dispatch: number; } & {}'.
      Type '{ store: string; }' is not assignable to type '{ dispatch: number; }'.
        Property 'dispatch' is missing in type '{ store: string; }'.
```

**Discussion**
I encountered the error while trying to correct the types for the `redux-react` module; consequently, `brokenFunction` represents a radically simplified version of that package's `connect` function.

The actual error message seems weird; the is no obvious requirement for the function passed in to expect to have `dispatch`.

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -7158,11 +7158,10 @@ namespace ts {
                             inferFromTypes(source, t);
                         }
                     }
-                    // Next, if target is a union type containing a single naked type parameter, make a
-                    // secondary inference to that type parameter. We don't do this for intersection types
-                    // because in a target type like Foo & T we don't know how which parts of the source type
-                    // should be matched by Foo and which should be inferred to T.
-                    if (target.flags & TypeFlags.Union && typeParameterCount === 1) {
+                    // Next, if target containings a single naked type parameter, make a secondary inference to that type
+                    // parameter. This gives meaningful results for union types in co-variant positions and intersection
+                    // types in contra-variant positions (such as callback parameters).
+                    if (typeParameterCount === 1) {
                         inferiority++;
                         inferFromTypes(source, typeParameter);
                         inferiority--;

--- a/tests/baselines/reference/intersectionTypeInference1.js
+++ b/tests/baselines/reference/intersectionTypeInference1.js
@@ -1,0 +1,17 @@
+//// [intersectionTypeInference1.ts]
+// Repro from #8801
+
+function alert(s: string) {}
+
+const parameterFn = (props:{store:string}) => alert(props.store)
+const brokenFunction = <OwnProps>(f: (p: {dispatch: number} & OwnProps) => void) => (o: OwnProps) => o
+export const Form3 = brokenFunction(parameterFn)({store: "hello"})
+
+
+//// [intersectionTypeInference1.js]
+// Repro from #8801
+"use strict";
+function alert(s) { }
+var parameterFn = function (props) { return alert(props.store); };
+var brokenFunction = function (f) { return function (o) { return o; }; };
+exports.Form3 = brokenFunction(parameterFn)({ store: "hello" });

--- a/tests/baselines/reference/intersectionTypeInference1.symbols
+++ b/tests/baselines/reference/intersectionTypeInference1.symbols
@@ -1,0 +1,33 @@
+=== tests/cases/compiler/intersectionTypeInference1.ts ===
+// Repro from #8801
+
+function alert(s: string) {}
+>alert : Symbol(alert, Decl(intersectionTypeInference1.ts, 0, 0))
+>s : Symbol(s, Decl(intersectionTypeInference1.ts, 2, 15))
+
+const parameterFn = (props:{store:string}) => alert(props.store)
+>parameterFn : Symbol(parameterFn, Decl(intersectionTypeInference1.ts, 4, 5))
+>props : Symbol(props, Decl(intersectionTypeInference1.ts, 4, 21))
+>store : Symbol(store, Decl(intersectionTypeInference1.ts, 4, 28))
+>alert : Symbol(alert, Decl(intersectionTypeInference1.ts, 0, 0))
+>props.store : Symbol(store, Decl(intersectionTypeInference1.ts, 4, 28))
+>props : Symbol(props, Decl(intersectionTypeInference1.ts, 4, 21))
+>store : Symbol(store, Decl(intersectionTypeInference1.ts, 4, 28))
+
+const brokenFunction = <OwnProps>(f: (p: {dispatch: number} & OwnProps) => void) => (o: OwnProps) => o
+>brokenFunction : Symbol(brokenFunction, Decl(intersectionTypeInference1.ts, 5, 5))
+>OwnProps : Symbol(OwnProps, Decl(intersectionTypeInference1.ts, 5, 24))
+>f : Symbol(f, Decl(intersectionTypeInference1.ts, 5, 34))
+>p : Symbol(p, Decl(intersectionTypeInference1.ts, 5, 38))
+>dispatch : Symbol(dispatch, Decl(intersectionTypeInference1.ts, 5, 42))
+>OwnProps : Symbol(OwnProps, Decl(intersectionTypeInference1.ts, 5, 24))
+>o : Symbol(o, Decl(intersectionTypeInference1.ts, 5, 85))
+>OwnProps : Symbol(OwnProps, Decl(intersectionTypeInference1.ts, 5, 24))
+>o : Symbol(o, Decl(intersectionTypeInference1.ts, 5, 85))
+
+export const Form3 = brokenFunction(parameterFn)({store: "hello"})
+>Form3 : Symbol(Form3, Decl(intersectionTypeInference1.ts, 6, 12))
+>brokenFunction : Symbol(brokenFunction, Decl(intersectionTypeInference1.ts, 5, 5))
+>parameterFn : Symbol(parameterFn, Decl(intersectionTypeInference1.ts, 4, 5))
+>store : Symbol(store, Decl(intersectionTypeInference1.ts, 6, 50))
+

--- a/tests/baselines/reference/intersectionTypeInference1.types
+++ b/tests/baselines/reference/intersectionTypeInference1.types
@@ -1,0 +1,41 @@
+=== tests/cases/compiler/intersectionTypeInference1.ts ===
+// Repro from #8801
+
+function alert(s: string) {}
+>alert : (s: string) => void
+>s : string
+
+const parameterFn = (props:{store:string}) => alert(props.store)
+>parameterFn : (props: { store: string; }) => void
+>(props:{store:string}) => alert(props.store) : (props: { store: string; }) => void
+>props : { store: string; }
+>store : string
+>alert(props.store) : void
+>alert : (s: string) => void
+>props.store : string
+>props : { store: string; }
+>store : string
+
+const brokenFunction = <OwnProps>(f: (p: {dispatch: number} & OwnProps) => void) => (o: OwnProps) => o
+>brokenFunction : <OwnProps>(f: (p: { dispatch: number; } & OwnProps) => void) => (o: OwnProps) => OwnProps
+><OwnProps>(f: (p: {dispatch: number} & OwnProps) => void) => (o: OwnProps) => o : <OwnProps>(f: (p: { dispatch: number; } & OwnProps) => void) => (o: OwnProps) => OwnProps
+>OwnProps : OwnProps
+>f : (p: { dispatch: number; } & OwnProps) => void
+>p : { dispatch: number; } & OwnProps
+>dispatch : number
+>OwnProps : OwnProps
+>(o: OwnProps) => o : (o: OwnProps) => OwnProps
+>o : OwnProps
+>OwnProps : OwnProps
+>o : OwnProps
+
+export const Form3 = brokenFunction(parameterFn)({store: "hello"})
+>Form3 : { store: string; }
+>brokenFunction(parameterFn)({store: "hello"}) : { store: string; }
+>brokenFunction(parameterFn) : (o: { store: string; }) => { store: string; }
+>brokenFunction : <OwnProps>(f: (p: { dispatch: number; } & OwnProps) => void) => (o: OwnProps) => OwnProps
+>parameterFn : (props: { store: string; }) => void
+>{store: "hello"} : { store: string; }
+>store : string
+>"hello" : string
+

--- a/tests/cases/compiler/intersectionTypeInference1.ts
+++ b/tests/cases/compiler/intersectionTypeInference1.ts
@@ -1,0 +1,7 @@
+// Repro from #8801
+
+function alert(s: string) {}
+
+const parameterFn = (props:{store:string}) => alert(props.store)
+const brokenFunction = <OwnProps>(f: (p: {dispatch: number} & OwnProps) => void) => (o: OwnProps) => o
+export const Form3 = brokenFunction(parameterFn)({store: "hello"})

